### PR TITLE
Terminate active sessions at privileges removal of the collaborator user

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Revoke the access token and terminate all active session, when the privileges of a collaborator user are removed.

Merge after following PR merged
- https://github.com/wso2/carbon-identity-framework/pull/3675
- https://github.com/wso2/carbon-identity-framework/pull/3693

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java [hunk 1]
```diff
@@ -86,6 +99,43 @@ public void handleEvent(Event event) throws IdentityEventException {
...
+                if (userIdList instanceof List<?>) {
+                    deletedUserIDList = (List<String>) userIdList;
+                    terminateSession(deletedUserIDList);
+                } else {
+                    String errorMsg = "Error occurred due to invalid format of deleted user Id list.";
+                    throw new IdentityEventException(errorMsg);
+                }
+            }
+
+        } else if (IdentityEventConstants.Event.PRE_DELETE_ROLE.equals(event.getEventName()) ||
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java:119]: let's revoke session on POST_DELETE_ROLE operation

Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java:119]: Have to go with the `PRE_DELETE_ROLE` as after role deletion, cannot retrieve the user Id list to revoke tokens and terminate sessions.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java [hunk 2]
```diff
@@ -115,4 +165,43 @@ private void revokeTokensOfDisabledUser(String userName, UserStoreManager userSt
...
+     * @throws IdentityEventException
+     */
+    private void terminateSession(List<String> userIDList) throws IdentityEventException {
+
+        try {
+            UserStoreManager userStoreManager = (UserStoreManager) CarbonContext.getThreadLocalCarbonContext()
+                    .getUserRealm().getUserStoreManager();
+
+            String userName;
+            if (!userIDList.isEmpty()) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java:182]: let's use CollectionUtils.isNotEmpty(userIDList)

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java [hunk 3]
```diff
@@ -86,6 +99,43 @@ public void handleEvent(Event event) throws IdentityEventException {
...
+
+            String roleId = (String) event.getEventProperties()
+                    .get(IdentityEventConstants.EventProperty.ROLE_ID);
+            String tenantDomain = (String) event.getEventProperties()
+                    .get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
+            try {
+                List<UserBasicInfo> userList = roleDAO.getRole(roleId, tenantDomain).getUsers();
+
+                List<String> userIdList = new ArrayList<>();
+                for (UserBasicInfo userBasicInfo : userList) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java:130]: add a null check if userList can become null here.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java]: Addressed with commit d038871c42d10fc171d32bc217d7cf75ebe2336e

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java]: Addressed with commit d038871c42d10fc171d32bc217d7cf75ebe2336e

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java [hunk 4]
```diff
@@ -86,6 +99,43 @@ public void handleEvent(Event event) throws IdentityEventException {
...
             }
+
+        } else if (IdentityEventConstants.Event.POST_UPDATE_USER_LIST_OF_ROLE.equals(event.getEventName())) {
+
+            Object userIdList = event.getEventProperties()
+                    .get(IdentityEventConstants.EventProperty.DELETE_USER_ID_LIST);
+            List<String> deletedUserIDList;
+
+            if (userIdList != null) {
+                if (userIdList instanceof List<?>) {
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOauthEventHandler.java]: Addressed with commit d038871c42d10fc171d32bc217d7cf75ebe2336e

**Review by emswbandara — APPROVED** [review]

**Pull request merged**